src/aurora/gff3writer.cpp

```diff
@@ -0,0 +1,491 @@
+	}
+
+	// Write field data.
+	for (size_t i = 0; i < _fields.size(); ++i) {
+		const Field &field = _fields[i];
+		switch (field.type) {
+			case GFF3Struct::kFieldTypeUint64:
+				stream.writeUint64LE(field.uint64Value);
+				break;
+			case GFF3Struct::kFieldTypeSint64:
+				stream.writeSint64LE(field.int64Value);
+				break;
+			case GFF3Struct::kFieldTypeDouble:
+				stream.writeIEEEDoubleLE(field.doubleValue);
+				break;
+			case GFF3Struct::kFieldTypeStrRef:
+				stream.writeUint32LE(4);
+				stream.writeUint32LE(field.uint32Value);
+				break;
+			case GFF3Struct::kFieldTypeResRef:
+				stream.writeByte(MIN(static_cast<byte>(16), static_cast<byte>(field.stringValue.size())));
+				stream.write(field.stringValue.c_str(), MIN<size_t>(field.stringValue.size(), 16));
+				break;
+			case GFF3Struct::kFieldTypeExoString:
+				stream.writeUint32LE(static_cast<uint32>(field.stringValue.size()));
+				stream.writeString(field.stringValue);
+				break;
+			case GFF3Struct::kFieldTypeLocString:
+				stream.writeUint32LE(field.locStringValue.getWrittenSize() + 8);
+				stream.writeUint32LE(field.locStringValue.getID());
+				stream.writeUint32LE(field.locStringValue.getNumStrings());
+				field.locStringValue.writeLocString(stream);
+				break;
+			case GFF3Struct::kFieldTypeVoid:
+				stream.writeUint32LE(static_cast<uint32>(field.voidSize));
+				stream.write(field.voidData.get(), field.voidSize);
+				break;
+			case GFF3Struct::kFieldTypeVector:
+				stream.writeIEEEFloatLE(field.vectorValue.x);
+				stream.writeIEEEFloatLE(field.vectorValue.y);
+				stream.writeIEEEFloatLE(field.vectorValue.z);
+				break;
+			case GFF3Struct::kFieldTypeOrientation:
+				stream.writeIEEEFloatLE(field.vectorValue.x);
+				stream.writeIEEEFloatLE(field.vectorValue.y);
+				stream.writeIEEEFloatLE(field.vectorValue.z);
+				stream.writeIEEEFloatLE(field.vectorValue.w);
+				break;
+			default:
+				break;
+		}
+	}
+
+	// Write field indices of every struct with more than one field.
+	for (size_t i = 0; i < _structs.size(); ++i) {
+		GFF3WriterStructPtr strct = _structs[i];
+		if (strct->getFieldCount() <= 1)
+			continue;
+
+		for (size_t j = 0; j < strct->getFieldCount(); ++j) {
+			stream.writeUint32LE(strct->_fieldIndices[j]);
+		}
+	}
+
+	// Write list indices.
+	for (size_t i = 0; i < _lists.size(); ++i) {
+		GFF3WriterListPtr list = _lists[i];
+		stream.writeUint32LE(list->getSize());
+		for (size_t j = 0; j < list->_strcts.size(); ++j) {
+			stream.writeUint32LE(list->_strcts[j]);
+		}
+	}
+}
+
+uint32 GFF3Writer::addLabel(const Common::UString &label) {
+	std::vector<Common::UString>::iterator iter = std::find(_labels.begin(), _labels.end(), label);
+	if (iter != _labels.end()) {
+		return static_cast<uint32>(std::distance(_labels.begin(), iter));
+	} else {
+		_labels.push_back(label);
+		return static_cast<uint32>(_labels.size() - 1);
+	}
+}
+
+uint32 GFF3Writer::getFieldDataSize(const Aurora::GFF3Writer::Field &field) {
```

Comment:
That can even be a static function in this file's scope. It doesn't need to be a method of GFF3Writer.